### PR TITLE
Add packed dataset

### PR DESCRIPTION
**Pull request by lantiga**
~Combined dataset testing is still TODO, the rest is ready for review.~

CombinedDataset will come in a later PR

### CODE REVIEW

File: lit_llama/packed_dataset.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+            else range(n_all_blocks)
+        )
+
+        self._curr_idx = 0
+
+    # def __getstate__(self):
+    #     # TODO: update
+    #     return self._filenames, self._n_chunks, self._seed
+
+    # def __setstate__(self, state):
```
Comment (awaelchli) [inline lit_llama/packed_dataset.py:111]: When would these be invoked?

**Review by awaelchli — APPROVED** [review]

File: lit_llama/packed_dataset.py (hunk 1 above)
Comment (awaelchli) [reply lit_llama/packed_dataset.py:111]: Was there some pickling going on?

File: lit_llama/packed_dataset.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        elem_id = (block_idx % self._n_blocks) * self._block_size
+        offset = np.dtype(self._dtype).itemsize * elem_id
+        arr = np.frombuffer(
+            buffer, dtype=self._dtype, count=self._block_size, offset=offset
+        )
+        self._curr_idx += 1
+        return arr
+
+
+class PackedDataset(IterableDataset):
```
[the remark this answers is not in the document]
Comment (lantiga) [reply lit_llama/packed_dataset.py:139]: Yes you're right

File: lit_llama/packed_dataset.py (hunk 1 above)
Comment (lantiga) [reply lit_llama/packed_dataset.py]: Eventually being able to serialize and deserialize the dataset deterministically is going to be important but I removed for now.

**Pull request merged**